File: token-procedure-archetype/src/main/resources/archetype-resources/src/main/java/__pluginName__/__pluginName__AuthorizeCodeTokenProcedure.java
package ${package}.${pluginName};

import se.curity.identityserver.sdk.data.tokens.TokenIssuerException;
import se.curity.identityserver.sdk.procedure.token.AuthorizeCodeTokenProcedure;
import se.curity.identityserver.sdk.procedure.token.context.AuthorizeTokenProcedurePluginContext;
import se.curity.identityserver.sdk.procedure.token.context.OpenIdConnectAuthorizeTokenProcedurePluginContext;
import se.curity.identityserver.sdk.service.issuer.NonceIssuer;
import se.curity.identityserver.sdk.web.ResponseModel;

import java.util.HashMap;

public final class ${pluginName}AuthorizeCodeTokenProcedure implements AuthorizeCodeTokenProcedure
{
    private final ${pluginName}TokenProcedureConfig _configuration;
    private final NonceIssuer authorizationCodeIssuer;

    public ${pluginName}AuthorizeCodeTokenProcedure(${pluginName}TokenProcedureConfig configuration)
    {
        _configuration = configuration;
        authorizationCodeIssuer = _configuration.getAuthorizationCodeIssuer();
    }

    @Override
    public ResponseModel run(AuthorizeTokenProcedurePluginContext context)
    {
        var authorizationCodeData = context.getDefaultAuthorizationCodeData();

        try
        {
            var issuedAuthorizationCode = authorizationCodeIssuer.issue(authorizationCodeData);

            var responseData = new HashMap<String, Object>(4);
            responseData.put("code", issuedAuthorizationCode);
            responseData.put("state", context.getProvidedState());
            responseData.put("iss", context.getIssuer());

            if (context.getScopeNames().contains("openid"))
            {
                if (context instanceof OpenIdConnectAuthorizeTokenProcedurePluginContext)
                {
                    responseData.put("session_state", ((OpenIdConnectAuthorizeTokenProcedurePluginContext) context).getSessionState());
                }
            }

            return ResponseModel.mapResponseModel(responseData);
        }
        catch (TokenIssuerException e)
        {
            return ResponseModel.problemResponseModel("token_issuer_exception", "Could not issue new tokens");
        }
    }
}

File: kotlin-token-procedure-archetype/src/main/resources/archetype-resources/src/main/kotlin/__pluginName__/__pluginName__AuthorizeCodeTokenProcedure.kt
package ${package}.${pluginName}

import se.curity.identityserver.sdk.data.tokens.TokenIssuerException
import se.curity.identityserver.sdk.procedure.token.AuthorizeCodeTokenProcedure
import se.curity.identityserver.sdk.procedure.token.context.AuthorizeTokenProcedurePluginContext
import se.curity.identityserver.sdk.procedure.token.context.OpenIdConnectAuthorizeTokenProcedurePluginContext
import se.curity.identityserver.sdk.web.ResponseModel

class ${pluginName}AuthorizeCodeTokenProcedure(private val _configuration: ${pluginName}TokenProcedureConfig): AuthorizeCodeTokenProcedure
{
    private val authorizationCodeIssuer = _configuration.getAuthorizationCodeIssuer()

    override fun run(context: AuthorizeTokenProcedurePluginContext): ResponseModel
    {
        val authorizationCodeData = context.defaultAuthorizationCodeData
        return try
        {
            val issuedAuthorizationCode = authorizationCodeIssuer.issue(authorizationCodeData)
            val responseData = mutableMapOf<String, Any>(
                "code" to issuedAuthorizationCode,
                "state" to context.providedState,
                "iss" to context.issuer
            )

            if (context.scopeNames.contains("openid") && context is OpenIdConnectAuthorizeTokenProcedurePluginContext)
            {
                responseData["session_state"] = context.sessionState
            }

            ResponseModel.mapResponseModel(responseData)
        }
        catch (e: TokenIssuerException)
        {
            ResponseModel.problemResponseModel("token_issuer_exception", "Could not issue new tokens")
        }
    }
}
